File: android/src/main/java/com/it_nomads/flutter_realm/SyncUserMethodSubHandler.java
package com.it_nomads.flutter_realm;


import java.util.ArrayList;
import java.util.Collections;
import java.util.HashMap;
import java.util.Map;

import io.flutter.plugin.common.MethodCall;
import io.flutter.plugin.common.MethodChannel;
import io.realm.ObjectServerError;
import io.realm.SyncCredentials;
import io.realm.SyncUser;

public class SyncUserMethodSubHandler implements MethodSubHandler {
    @Override
    public boolean onMethodCall(MethodCall call, MethodChannel.Result result) {
        switch (call.method) {
            case "logInWithCredentials":
                onLogInWithCredentials((Map) call.arguments, result);
                return true;
            case "allUsers":
                onAllUsers(result);
                return true;
            case "currentUser":
                onCurrentUser(result);
                return true;
            case "logOut":
                onLogOut((Map) call.arguments, result);
                return true;
            default:
                return false;
        }
    }

    private void onLogOut(Map arguments, MethodChannel.Result result) {
        String identity = (String) arguments.get("identity");
        SyncUser user = SyncUser.all().get(identity);

        if (user == null) {
            result.error("User with identity = \"" + identity + "\" not found.", null, null);
        } else {
            user.logOut();
            result.success(null);
        }
    }

    private void onCurrentUser(MethodChannel.Result result) {
        SyncUser syncUser = SyncUser.current();

        if (syncUser == null) {
            result.success(null);
        } else {
            result.success(Collections.unmodifiableMap(userToMap(syncUser)));

        }
    }

    private void onAllUsers(MethodChannel.Result result) {
        ArrayList<Map> data = new ArrayList<>();

        for (SyncUser user : SyncUser.all().values()) {
            data.add(userToMap(user));
        }

        result.success(data);
    }

    private void onLogInWithCredentials(Map arguments, final MethodChannel.Result result) {
        SyncCredentials credentials = credentialsFromArguments(arguments);
        if (credentials == null) {
            String message = "Provider is not supported for authorization. Received: " + arguments.toString();
            result.error(message, null, null);
            return;
        }

        String url = (String) arguments.get("authServerURL");
        assert url != null;

        SyncUser.logInAsync(credentials, url, new SyncUser.Callback<SyncUser>() {
            @Override
            public void onSuccess(SyncUser user) {
                Map data = userToMap(user);
                result.success(data);
            }

            @Override
            public void onError(ObjectServerError error) {
                result.error(error.getErrorMessage(), null, null);
            }
        });

    }


    private Map userToMap(SyncUser user) {
        HashMap<String, Object> data = new HashMap<>();
        data.put("identity", user.getIdentity());
        data.put("isAdmin", user.isAdmin());

        return data;
    }

    private SyncCredentials credentialsFromArguments(Map arguments) {
        String provider = String.valueOf(arguments.get("provider"));
        Map data = (Map) arguments.get("data");
        assert data != null;

        if ("jwt".equals(provider)) {
            String jwt = (String) data.get("jwt");
            assert jwt != null;

            return SyncCredentials.jwt(jwt);
        }

        if ("username&password".equals(provider)) {
            String username = (String) data.get("username");
            String password = (String) data.get("password");
            Boolean shouldRegister = (Boolean) data.get("shouldRegister");

            return SyncCredentials.usernamePassword(username, password, shouldRegister);
        }
        return null;
    }
}

File: ios/Classes/SyncUserMethodSubHandler.m
//
//  FlutterSyncUser.m
//  flutter_realm
//
//  Created by German Saprykin on 9/8/19.
//

#import <Realm/Realm.h>
#import "SyncUserMethodSubHandler.h"
#import "FlutterError+NSError.h"


@interface SyncUserMethodSubHandler()
@end

@implementation SyncUserMethodSubHandler

- (BOOL)handleMethodCall:(FlutterMethodCall *)call result:(FlutterResult)result {
    NSString *aSelector = [NSString stringWithFormat:@"%@:result:", call.method];
    SEL selector = NSSelectorFromString(aSelector);
    if ([self respondsToSelector:selector]){
        [self performSelector:selector withObject:call.arguments withObject:result];
        return true;
    }
    return false;
}


- (void)logInWithCredentials:(NSDictionary *)arguments result:(FlutterResult)result {
    NSString *urlString = arguments[@"authServerURL"];
    NSURL *url = [NSURL URLWithString:urlString];
    
    RLMSyncCredentials *creds = [self credentialsFromArguments:arguments];
    if (creds == nil){
        NSString *message = [NSString stringWithFormat:@"Provider is nit supported for authorization. Received: %@", arguments];
        FlutterError *error =[FlutterError errorWithCode:@"-1" message:message details:nil];
        result(error);
        return;
    }
    
    [RLMSyncUser logInWithCredentials:creds authServerURL:url onCompletion:^(RLMSyncUser * _Nullable user, NSError * _Nullable error) {
        if (user != nil){
            NSMutableDictionary *map = [NSMutableDictionary dictionary];
            map[@"identity"] = user.identity;
            result(map);
        }else{
            result([FlutterError fromNSError:error]);
        }
    }];
}

- (void)allUsers:(NSDictionary *)arguments result:(FlutterResult)result {
    NSMutableArray *data = [NSMutableArray array];
    
    for (RLMSyncUser *user in [[RLMSyncUser allUsers] allValues]) {
        [data addObject:[self userToMap:user]];
    }
    result(data);
}

- (void)currentUser:(NSDictionary *)arguments result:(FlutterResult)result {
    RLMSyncUser *user = [RLMSyncUser currentUser];
    result([self userToMap:user]);
}


- (void)logOut:(NSDictionary *)arguments result:(FlutterResult)result {
    NSString *identity = arguments[@"identity"];
    RLMSyncUser *user = [RLMSyncUser allUsers][identity];
    if (user == nil) {
        NSString *message = [NSString stringWithFormat:@"User with identity \"%@\" is nil.", identity];
        FlutterError *error =[FlutterError errorWithCode:@"-1" message:message details:nil];
        result(error);
        return;
    }
    
    [user logOut];
    result(nil);
}

- (RLMSyncCredentials *)credentialsFromArguments:(NSDictionary *)arguments {
    NSString *provider = arguments[@"provider"];
    NSDictionary *data = arguments[@"data"];
    
    
    if ([@"jwt" isEqualToString:provider]){
        NSString *jwt = data[@"jwt"];
        return [RLMSyncCredentials credentialsWithJWT:jwt];
    }
    
    if ([@"username&password" isEqualToString:provider]) {
        NSString *email = data[@"username"];
        NSString *password = data[@"password"];
        BOOL shouldRegister = [data[@"shouldRegister"] boolValue];
        
        return [RLMSyncCredentials credentialsWithUsername:email password:password register:shouldRegister];
    }
    
    return nil;
}

- (NSDictionary *)userToMap:(RLMSyncUser *)user {
    NSMutableDictionary *map = [NSMutableDictionary dictionary];
    
    map[@"identity"] = [user identity];
    map[@"isAdmin"] = @([user isAdmin]);
    map[@"refreshToken"] = [user refreshToken];
    
    return map;
}


@end
